### ISSUE TITLE
Clippy incorrectly says that an infinite loop will never loop.

### ISSUE THREAD

**Issue by ChaiTRex** [rust-lang/rust-clippy#11004]
### Summary

When a macro is involved, Clippy incorrectly says that a loop will never loop, but the loop is an infinite loop, so it loops quite a bit.

### Lint Name

never_loop

### Reproducer

I [tried this code](https://play.rust-lang.org/?edition=2021&gist=acffef0d02905565635e63292922b4f4) (which is reduced from [a prime iterator](https://play.rust-lang.org/?edition=2021&gist=5ae227468ea4b0b1c56cbf5e13846eec)):

```rust
#[allow(dead_code)]
fn main() {
    macro_rules! m {
        () => {
            'label: {
                if true {
                    break 'label;
                }
                return;
            }
        };
    }

    loop {
        m!();
    }
}
```

I saw this happen:

```
$ cargo clippy
    Checking never_loop v0.1.0 (/home/chaitrex/dev/never_loop)
error: this loop never actually loops
  --> src/main.rs:14:5
   |
14 | /     loop {
15 | |         m!();
16 | |     }
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#never_loop
   = note: `#[deny(clippy::never_loop)]` on by default

error: could not compile `never_loop` (bin "never_loop") due to previous error
```

I expected to see this happen:

No errors or warnings from Clippy.

### Version

```text
rustc 1.70.0 (90c541806 2023-05-31)
binary: rustc
commit-hash: 90c541806f23a127002de5b4038be731ba1458ca
commit-date: 2023-05-31
host: x86_64-unknown-linux-gnu
release: 1.70.0
LLVM version: 16.0.2
```

**Comment by blyxyas** [comment]
Expanding this macro will also trigger this lint:

```
#[allow(dead_code)]
fn main() {
    loop {
        'label: {
            if true {
                break 'label;
            }
            return;
        }
    }
}
```

**Comment by Centri3** [comment]
It looks like it doesn't actually evaluate the `if` condition, so it'll just assume it always fails (I guess? More accurately, it never short-circuits if it's always true), maybe for simple expressions like `true` this can be fixed but something more complex is out of scope of clippy

**Comment by ChaiTRex** [comment]
Could it check if there's a `break` (or `continue`, etc.) anywhere before the `return` that would break out past the `return` or continue the `loop`?

**Comment by digama0** [comment]
I also ran across this:
```rust
fn main() {
    let mut i = 0;
    'a: loop {
        i += 1;
        println!("{i}");
        if i == 10 {
            break
        }
        'b: {
            for j in 0..2 {
                if j == 1 {
                    break 'b
                }
            }
            break 'a
        }
    }
}
```
There is no need to be evaluating the truthiness of conditionals here. This is an underapproximation gone bad - if you don't know whether an if statement evaluates to true or false you should assume both branches are reachable.

**Comment by digama0** [comment]
Looking at the code, I see it has some flawed reachability logic. Here's another FP (which loops infinitely when `b` is true):
```rust
fn foo(b: bool) {
    'a: loop {
        'b: {
            'c: loop {
                if b { break 'c } else { break 'b }
            }
            continue 'a
        }
        break 'a
    }
}
```
I will see if I can write a patch, although it would be my first contribution to clippy so hopefully it works out.
